packages/rx_bloc_cli/mason_templates/bricks/feature_cicd_fastlane/__brick__/{{#cicd_codemagic}}codemagic.yaml{{/cicd_codemagic}}

```diff
@@ -0,0 +1,111 @@
+definitions:
+  instance_mac_pro: &instance_mac_pro
+    instance_type: mac_mini_m1
+    max_build_duration: 100
+  environment: &environment
+    groups:
+      - secret
+    flutter: stable
+    xcode: latest
+    cocoapods: default
+  triggers: &triggers
+    events:
+      - tag
+    tag_patterns:
+      - pattern: '*-v[0-9]+.[0-9]+.[0-9]+\+[0-9]+'
+        include: true
+  cached: &cached
```

Comment:
If there is possible to cache flutter or other time/build minutes consuming tasks I would suggest to have a separate build steps such as

- build, which will produce artifacts (output)
- deploy, which will use the output from the previous step as a input and will upload artifacts 

Please research whether this is possible in the context of CodeMagic CD

Reply:
@StanevPrime Not sure what you mean about having separate build steps based on the current code. The current code does have separate steps for ios/android builds and is inspired by the github implementation. The only difference between codemagic and github implementation is that codemagic implementation needs to be within a single `codemagic.yaml` file (codemagic can use references to reusable definitions) while with github we can split the implementation between files (templates + actual workflow using the templates).

![Screenshot 2024-06-12 at 15 44 09](https://github.com/Prime-Holding/rx_bloc/assets/61236867/0fdd1e31-808c-4995-a108-22e5d6cd8b30)

---

As for the caching, I'm currently looking into whether or not we can optimise things even more. 🙌🏼

Reply:
As per [this documentation](https://docs.codemagic.io/knowledge-codemagic/caching/), the cache will be kept for 14 days and if needed it can be manually cleared. Each workflow (both android and ios workflows) [has it's own cache](https://docs.codemagic.io/yaml-basic-configuration/yaml-getting-started/?utm_source=app&utm_campaign=yaml-cache#cache).

Also, taking into consideration the following note:
![Screenshot 2024-06-12 at 16 18 07](https://github.com/Prime-Holding/rx_bloc/assets/61236867/3c5e6ff7-809a-4ff6-a700-2aa720ea21aa)

I'm assuming that each flutter version when downloaded is cached on the machine (downloading new ones which are missing). So I am puzzled whether or not to add the flutter path to cached ones. < @StanevPrime

Reply:
I've ran some tests for both android and iOS and it seems that caching flutter does help but with a price of a higher cache. One thing to have in mind is that the flutter cache might be growing, as time goes on. Another thing to consider caching flutter is that for a personal account the [max cache limit is 3GB per workflow](https://docs.codemagic.io/knowledge-codemagic/caching/#cache-usage-limits).

---

Below you can see a list of different scenarios for different cached paths and the used cache. Also, I've included a table of times it takes the whole jobs (and also the app builds) for iOS/Android for different scenarios. This has been tested on the existing project we used for setting up CI/CD for Github.

> [!NOTE]
> Same scenarios may have different sizes due to some cached paths not being in use for a given platform.

Scenario 1:
```
- $HOME/.gradle/caches
- $HOME/Library/Caches/CocoaPods
- $FCI_BUILD_DIR/ios/Pods
- $FCI_BUILD_DIR/android/.gradle
- $FLUTTER_ROOT
```

Scenario 2:
```
- $HOME/.gradle/caches
- $HOME/Library/Caches/CocoaPods
- $FCI_BUILD_DIR/ios/Pods
- $FCI_BUILD_DIR/android/.gradle
```

Cache size (GB):

Scenario | Android  |  iOS
---------|----------|-------
   1     |   2.71   |  1.00
   2     |   1.78   |  0.07

Job time / app build (seconds):

Scenario          |    Android    |  iOS
------------------|---------------|-------------
1 (without cache) |  575 / 263    |  546 / 383    
1 (with cache)    |  318 / 158    |  410 / 252    
2 (without cache) |  587 / 307    |  442 / 262    
2 (with cache)    |  533 / 296    |  486 / 287

Reply:
Do we have any update here? Can we build deploy-only workflow ?

Reply:
Here's the look of the workflow, accepting two inputs for deployment (the codemagic links to the artifact (`ipa` or `aab) and the `deployment.yaml` file). This was the nearest thing that I could think of according to our discussion about how to approach this.

![Screenshot 2024-06-18 at 15 34 42](https://github.com/Prime-Holding/rx_bloc/assets/61236867/5e3db7c8-dc56-4e60-9fe8-bc761ae87cc5)

Reply:
What's the purpose of selecting the branch, considering the artefacts are already produced?

Reply:
The fields `Select branch` and `Select file workflow` (under `Build branch`) are non-toggleable options provided by Codemagic.

Currently there are three workflows:
- Build Android app
- Build iOS app
- Deploy app

`Build Android app` and `Build iOS app` are workflows intended to be ran from the `Build tag` option, as they depend on the tag name to get the build details (such as the `flavor`,`build version` and `build number`). When starting a build from the `Build tag` option, `codemagic input`s (fields such as the `Build artifact` and `The deployment.yaml file`) aren't available for use.

The `Deploy app` is intended to be used from the `Build branch` option, allowing user input for the builds. We need those inputs (`Build artifact` and `The deployment.yaml file`) so we can retrieve the artifacts.

Reply:
In contrast, here's how the `Build tag` screen looks like with the same workflow:

![Screenshot 2024-06-18 at 17 39 29](https://github.com/Prime-Holding/rx_bloc/assets/61236867/3ec351f6-3750-4a2a-8d99-5ca3f7b7b0a4)